trie/triedb/pathdb/nodebufferlist.go

```diff
@@ -73,6 +74,129 @@ type nodebufferlist struct {
 	forceKeepCh     chan struct{}  // Trigger force keep event loop.
 	waitForceKeepCh chan struct{}  // Wait force keep event loop.
 	keepFunc        NotifyKeepFunc // Used to keep op-proposer output proof.
+
+	freezer  *rawdb.ResettableFreezer // Freezer for storing trie histories, nil possible in tests
+	recovery bool                     // Signals that pathdb is recovering diff layers
+}
+
+// newNodeBufferListForRecovery initializes the node buffer list for recovering diff layers
+func newNodeBufferListForRecovery(db ethdb.Database, freezer *rawdb.ResettableFreezer, base *multiDifflayer,
+	limit, wpBlocks, rsevMdNum, dlInMd uint64) *nodebufferlist {
+	nbl := &nodebufferlist{
+		db:              db,
+		wpBlocks:        wpBlocks,
+		rsevMdNum:       rsevMdNum,
+		dlInMd:          dlInMd,
+		limit:           limit,
+		base:            base,
+		stopCh:          make(chan struct{}),
+		waitStopCh:      make(chan struct{}),
+		forceKeepCh:     make(chan struct{}),
+		waitForceKeepCh: make(chan struct{}),
+		freezer:         freezer,
+		recovery:        true,
+	}
+	freezerLength, err := freezer.Ancients()
+	if err != nil {
+		log.Crit("Failed to get freezer ancients", "error", err)
+		return nil
+	}
+	tail, err := freezer.Tail()
+	if err != nil {
+		log.Crit("Failed to get freezer tail", "error", err)
+		return nil
+	}
+	persistentStateID := rawdb.ReadPersistentStateID(db)
+
+	log.Info("Print ancient db meta", "persistent state id", persistentStateID, "freezer length", freezerLength,
+		"freezer tail", tail, "waitingRecoverNum", freezerLength-persistentStateID)
+
+	var (
+		wg    sync.WaitGroup
+		start = time.Now()
+	)
+	intervals := nbl.createIntervalBoundaries(freezerLength, persistentStateID+1)
+	nbl.linkMultiDiffLayers(persistentStateID, freezerLength)
+	for current, i := nbl.head, 0; current != nil; current, i = current.next, i+1 {
+		wg.Add(1)
+		go func(index int, slice []uint64, mdl *multiDifflayer) {
+			defer wg.Done()
+			for j := slice[0]; j <= slice[1]; j++ {
+				h := nbl.readStateHistory(j)
+				if h == nil {
+					return
+				}
+				if err = mdl.commit(h.meta.root, j, h.meta.block, 1, flattenTrieNodes(h.nodes)); err != nil {
+					log.Crit("Failed to commit trie nodes to multi diff layer", "error", err)
+				}
+				// log.Info("concurrent info", "start", slice[0], "end", slice[1], "index", index)
+			}
+		}(i, intervals[i], current)
+	}
+	wg.Wait()
+
+	log.Info("diff to base", "tail state id", nbl.tail.id, "head", nbl.head.id)
+	nbl.diffToBase()
+	if nbl.base.size >= nbl.base.limit {
+		log.Info("background flush")
+		nbl.backgroundFlush()
```

Comment:
Just recover memory and merge, no disk operation. Don' flush as it maybe crush during recovery phase.

Reply:
Deleted